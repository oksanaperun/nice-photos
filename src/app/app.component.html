<h1>{{ title }}</h1>
<app-page></app-page>
