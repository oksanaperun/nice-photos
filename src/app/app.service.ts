import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs/Observable';
import { environment } from '../environments/environment';

@Injectable()
export class AppService {
  constructor(private http: HttpClient) { }

  searchItemsBySearchText(searchText): Observable<SearchResponse> {
    const header = new HttpHeaders().set('Accept-Version', 'v1');
    let params = new HttpParams().set('query', searchText);
    params = params.append('client_id', environment.clientId);
    params = params.append('per_page', '9');
    params = params.append('orientation', 'landscape');

    return this.http.get<SearchResponse>(`${environment.apiBase}/search/photos`, {
      headers: header,
      params: params
    });
  }
}

export interface SearchResponse {
  total: number;
  results: SearchResponseResult[];
}

export interface SearchResponseResult {
  id: string;
  urls: SearchResponseResultUrls;
}

export interface SearchResponseResultUrls {
  small: string;
  regular: string;
}

export interface SearchResultsData {
  totalCount: number;
  items: Item[];
}

export interface Item {
  id: string;
  smallUrl: string;
}
