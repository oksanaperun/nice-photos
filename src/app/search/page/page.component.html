<app-search-form (formSubmit)="onFormSubmit($event)"></app-search-form>
<app-search-results *ngIf="totalCount$ | async as totalCount" [totalCount]="totalCount" [items]="items"></app-search-results>
<app-search-spinner *ngIf="isLoading"></app-search-spinner>
<app-search-error *ngIf="isFailed"></app-search-error>
