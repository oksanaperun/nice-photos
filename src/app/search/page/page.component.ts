import { Component, ViewChild } from '@angular/core';
import { SearchComponent } from '../search';

@Component({
  selector: 'app-page',
  templateUrl: './page.component.html',
})

export class PageComponent {
  @ViewChild('search') searchComponent: SearchComponent;

  get isSpinner(): boolean {
    return this.searchComponent.isSearchStarted && !this.searchComponent.isSearchFinished;
  }

  get isResultsDataReady(): boolean {
    return this.searchComponent.isSearchFinished && !this.searchComponent.isError;
  }

  get isError(): boolean {
    return this.searchComponent.isSearchFinished && this.searchComponent.isError;
  }
}
