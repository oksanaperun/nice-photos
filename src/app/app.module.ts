import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { AppComponent } from './app.component';
import { SearchModule } from './search';
import { AppService } from './app.service';
import { WINDOW_PROVIDERS } from './window.service';

@NgModule({
  declarations: [AppComponent],
  imports: [
    BrowserModule,
    SearchModule,
  ],
  providers: [
    AppService,
    WINDOW_PROVIDERS,
  ],
  bootstrap: [AppComponent]
})

export class AppModule { }
