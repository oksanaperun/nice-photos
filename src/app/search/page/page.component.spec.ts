import { TestBed, ComponentFixture } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { By } from '@angular/platform-browser';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs/Observable';
import 'rxjs/add/observable/of';
import 'rxjs/add/observable/throw';
import 'rxjs/add/operator/delay';
import { PageComponent } from './page.component';
import { AppService, SearchResponse, SearchResponseResult, Item } from '../../app.service';

describe('PageComponent', () => {
  let component: PageComponent;
  let fixture: ComponentFixture<PageComponent>;
  let appService: AppService;
  const searchText = 'abc';
  const result: SearchResponseResult = {
    id: 'some-id',
    color: '#add4dd',
    urls: { small: 'some-url', regular: 'some-url' }
  };
  const transformedResult: Item = {
    id: result.id,
    color: '#add4dd',
    smallUrl: result.urls.small
  };

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [PageComponent],
      providers: [AppService, { provide: HttpClient, useValue: null }],
      schemas: [NO_ERRORS_SCHEMA],
    }).compileComponents();

    fixture = TestBed.createComponent(PageComponent);
    component = fixture.componentInstance;
  });

  it('should render only search form by default', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should render spinner when search started', () => {
    component.isLoading = true;
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should render search results when there are search results data', () => {
    component.totalCount = 0;
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should render error when search finished with error', () => {
    component.isFailed = true;
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should call #searchItems on form submit', () => {
    spyOn(component, 'searchItems');
    component.onFormSubmit(searchText);

    expect(component.searchItems).toHaveBeenCalledWith(searchText);
  });

  it('should set properties on loading when search items', () => {
    appService = fixture.debugElement.injector.get(AppService);
    spyOn(appService, 'getItems').and.returnValue(Observable.of(null).delay(200));

    component.searchItems(searchText);

    component.items$.subscribe(() => {
      expect(component.isLoading).toBe(true);
      expect(component.isFailed).toBe(false);
    });
  });

  it('should call AppService with search text and page number when search items', () => {
    appService = fixture.debugElement.injector.get(AppService);
    spyOn(appService, 'getItems').and.returnValue(Observable.of(null));

    component.searchItems(searchText);

    component.items$.subscribe(() => {
      expect(appService.getItems).toHaveBeenCalledWith(searchText, 1);
    });
  });

  it('should set properties on success call to AppService when search items', () => {
    const response: SearchResponse = { total: 2, total_pages: 3, results: [result, result] };

    appService = fixture.debugElement.injector.get(AppService);
    spyOn(appService, 'getItems').and.returnValue(Observable.of(response));

    component.searchItems(searchText);

    component.items$.subscribe(response => {
      expect(component.isLoading).toBe(false);
      expect(component.isFailed).toBe(false);
      expect(component.totalCount).toBe(2);
      expect(component.totalPagesNumber).toBe(3);
      expect(response).toEqual([transformedResult, transformedResult]);
    })
  });

  it('should set properties on failed call to AppService when search items', () => {
    appService = fixture.debugElement.injector.get(AppService);
    spyOn(appService, 'getItems').and.returnValue(Observable.throw({ status: 500 }));

    component.searchItems(searchText);

    component.items$.subscribe(response => {
      expect(component.isLoading).toBe(false);
      expect(component.isFailed).toBe(true);
      expect(response).toBe(null);
    });
  });
});
