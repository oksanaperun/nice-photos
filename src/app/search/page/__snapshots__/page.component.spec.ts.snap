// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageComponent should render error when search finished with error 1`] = `
<app-page
  appService={[Function AppService]}
  debounceTime={[Function Number]}
  isFailed={[Function Boolean]}
  itemHeight={[Function Number]}
  numberOfItems={[Function Number]}
  pageByManual$={[Function BehaviorSubject]}
  pageByResize$={[Function Observable]}
  pageByScroll$={[Function Observable]}
  pageToLoad$={[Function Observable]}
>
  <app-search-form />
  
  
  <app-search-error />
  
</app-page>
`;

exports[`PageComponent should render only search form by default 1`] = `
<app-page
  appService={[Function AppService]}
  debounceTime={[Function Number]}
  itemHeight={[Function Number]}
  numberOfItems={[Function Number]}
  pageByManual$={[Function BehaviorSubject]}
  pageByResize$={[Function Observable]}
  pageByScroll$={[Function Observable]}
  pageToLoad$={[Function Observable]}
>
  <app-search-form />
  
  
  
  
</app-page>
`;

exports[`PageComponent should render search results when there are search results data 1`] = `
<app-page
  appService={[Function AppService]}
  debounceTime={[Function Number]}
  itemHeight={[Function Number]}
  numberOfItems={[Function Number]}
  pageByManual$={[Function BehaviorSubject]}
  pageByResize$={[Function Observable]}
  pageByScroll$={[Function Observable]}
  pageToLoad$={[Function Observable]}
  totalCount="0"
>
  <app-search-form />
  
  
  
  
</app-page>
`;

exports[`PageComponent should render spinner when search started 1`] = `
<app-page
  appService={[Function AppService]}
  debounceTime={[Function Number]}
  isLoading={[Function Boolean]}
  itemHeight={[Function Number]}
  numberOfItems={[Function Number]}
  pageByManual$={[Function BehaviorSubject]}
  pageByResize$={[Function Observable]}
  pageByScroll$={[Function Observable]}
  pageToLoad$={[Function Observable]}
>
  <app-search-form />
  
  <app-search-spinner />
  
  
</app-page>
`;
